File: burstLibrary/src/main/java/com/harry1453/burst/explorer/service/NetworkObjectService.java
package com.harry1453.burst.explorer.service;

import io.reactivex.Single;

public final class NetworkObjectService implements ObjectService {

    private final NetworkService networkService;
    private final DeserializerService deserializerService;

    NetworkObjectService(NetworkService networkService, DeserializerService deserializerService) {
        this.networkService = networkService;
        this.deserializerService = deserializerService;
    }

    @Override
    public <T> Single<T> fetchObject(String url, Class<T> objectClass) {
        return networkService.fetchData(url)
                .flatMap(objectData -> deserializerService.deserializeObject(objectData, objectClass));
    }
}

File: burstLibrary/src/main/java/com/harry1453/burst/explorer/service/NetworkObjectService.kt
package com.harry1453.burst.explorer.service

import io.reactivex.Single

class NetworkObjectService internal constructor(private val networkService: NetworkService, private val deserializerService: DeserializerService) : ObjectService {
    override fun <T> fetchObject(url: String, objectClass: Class<T>): Single<T> {
        return networkService.fetchData(url)
                .flatMap { objectData -> deserializerService.deserializeObject(objectData, objectClass) }
    }
}
